src/SonarScanner.MSBuild.PreProcessor/JreCaching/JreCache.cs

```diff
@@ -45,6 +47,108 @@ public JreCacheResult IsJreCached(string sonarUserHome, JreDescriptor jreDescrip
         return new JreCacheFailure($"The JRE cache directory in '{Path.Combine(sonarUserHome, "cache")}' could not be created.");
     }
 
+    public async Task<JreCacheResult> DownloadJreAsync(string sonarUserHome, JreDescriptor jreDescriptor, Func<Task<Stream>> jreDownload)
+    {
+        if (EnsureCacheRoot(sonarUserHome, out var cacheRootLocation)
+            && EnsureDirectoryExists(Path.Combine(cacheRootLocation, jreDescriptor.Sha256)) is { } jreDownloadPath)
+        {
+            var downloadTarget = Path.Combine(jreDownloadPath, jreDescriptor.Filename);
+            return fileWrapper.Exists(downloadTarget)
+                ? await UnpackJre(downloadTarget, jreDescriptor, cacheRootLocation)
+                : await DownloadAndUnpackJre(jreDownloadPath, downloadTarget, jreDescriptor, cacheRootLocation, jreDownload);
+        }
+        else
+        {
+            return new JreCacheFailure($"The JRE cache directory in '{Path.Combine(sonarUserHome, "cache", jreDescriptor.Sha256)}' could not be created.");
+        }
+    }
+
+    private async Task<JreCacheResult> DownloadAndUnpackJre(string jreDownloadPath, string downloadTarget, JreDescriptor jreDescriptor, string cacheRootLocation, Func<Task<Stream>> jreDownload)
+    {
+        if (await DownloadJre(jreDownloadPath, downloadTarget, jreDownload) is { } exception)
+        {
+            if (fileWrapper.Exists(downloadTarget)) // Even though the download failed, there is a small chance the file was downloaded by another scanner in the meantime.
+            {
+                return await UnpackJre(downloadTarget, jreDescriptor, cacheRootLocation);
+            }
+            return new JreCacheFailure(string.Format(Resources.ERR_JreDownloadFailed, exception.Message));
+        }
+        else
+        {
+            return await UnpackJre(downloadTarget, jreDescriptor, cacheRootLocation);
+        }
+    }
+
+    private async Task<Exception> DownloadJre(string jreDownloadPath, string downloadTarget, Func<Task<Stream>> jreDownload)
+    {
+        // We download to a temporary file in the right folder.
+        // This avoids conflicts, if multiple scanner try to download to the same file.
+        var tempFileName = Path.GetRandomFileName();
+        var tempFile = Path.Combine(jreDownloadPath, tempFileName);
+        try
+        {
+            using var fileStream = fileWrapper.Create(tempFile);
```

Comment:
I resolved the other issue.
My main problem is not the functionality, but the readability.
I think being extra-safe is very fair, but maybe you could re-write it with local functions so that we semantically know what is what:


```csharp

        // We download to a temporary file in the right folder.
        // This avoids conflicts, if multiple scanner try to download to the same file.
        var tempFileName = Path.GetRandomFileName();
        var tempFile = Path.Combine(jreDownloadPath, tempFileName);
        try
        {
            using var fileStream = fileWrapper.Create(tempFile);
            return await DownloadAndMove(fileStream);
        }
        catch (Exception ex)
        {
            return ex;
        }

        async Task<Exception> DownloadAndMove(Stream tempStream)
        {
            try
            {
                using var downloadStream = await jreDownload();
                await downloadStream.CopyToAsync(tempStream);
                tempStream.Close();
                fileWrapper.Move(tempFile, downloadTarget);
                return null;
            }
            catch
            {
                Cleanup(tempStream);
                throw;
            }
        }

        void Cleanup(Stream tempStream)
        {
            try
            {
                // Cleanup the temp file
                EnsureClosed(tempStream); // If we do not close  the stream, deleting the file fails with:
                                          // The process cannot access the file '<<path-to-file>>' because it is being used by another process.
                fileWrapper.Delete(tempFile);
            }
            catch
            {
                // Ignore any failures to delete the temp file
            }
        }
```

Reply:
I do not see how this is any better readable. It is just different. The flow is complicated and no matter how you change it, it will always be complicated. Given that, a single big method is better suited to this situation, in my opinion. Also, your code is a mixture of captured and passed variables and, therefore, is hard to read. Also, it doesn't play well with the logging we will need to add (see #2015).